### PR TITLE
`create_protostar_project` fixture

### PR DESCRIPTION
**Pull request by kasperski95**
- added `create_protostar_project`
- refactored `protostar format` integration tests
---
- added deploy command to `ProtostarFixture` and used it in `tests/integration/deploying`

### CODE REVIEW

File: tests/integration/conftest.py [hunk 1]
```diff
@@ -163,7 +165,35 @@ def protostar_fixture(
     session_mocker: MockerFixture,
     protostar_project_root_path: Path,
 ) -> ProtostarFixture:
+    """@deprecated: Use `create_protostar_project` fixture instead."""
```
Comment (Arcticae) [inline tests/integration/conftest.py:168]: why not replace all usages right now?

File: tests/integration/conftest.py [hunk 2]
```diff
@@ -163,7 +165,35 @@ def protostar_fixture(
...
     return build_protostar_fixture(
         mocker=session_mocker,
         project_root_path=protostar_project_root_path,
     )
+
+
+class CreateProtostarProjectFixture(Protocol):
```
Comment (Arcticae) [inline tests/integration/conftest.py:175]: I don't understand how this is helping us, is that something that we really need?  Typing the fixtures seems a little excessive

File: tests/integration/conftest.py [hunk 3]
```diff
@@ -163,7 +165,35 @@ def protostar_fixture(
...
+
+ProtostarProjectFixture = ProtostarFixture
+
+
+@pytest.fixture(name="create_protostar_project", scope="module")
+def create_protostar_project_fixture(
+    session_mocker: MockerFixture,
+):
+    @contextmanager
+    def create_protostar_project(project_root_path: Path):
```
Comment (Arcticae) [inline tests/integration/conftest.py:188]: This function returns protostar fixture instance, but you're naming the variables `protostar_project`, this seems misleading to what this really returns

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: tests/integration/conftest.py (hunk 2 above)
Comment (mkaput) [reply tests/integration/conftest.py:175]: it helps IDE autocompletion

**Review by mkaput — CHANGES_REQUESTED** [review]
To my eyes, usage of this fixture looks like still invocations of protostar and personally I would drop the `project` word from the name, i.e. replace protostar_fixture with this. Is something blocking such move?

File: tests/integration/conftest.py (hunk 2 above)
Comment (kasperski95) [reply tests/integration/conftest.py:175]: And refactoring, and gives immediate feedback in case of misuse, and [insert type checking benefits]

**Comment by kasperski95** [comment]
> To my eyes, usage of this fixture looks like still invocations of protostar and personally I would drop the project word from the name, i.e. replace protostar_fixture with this. Is something blocking such move?

- conflict with the `protostar` fixture (which I'll replace in this PR)
- `ProtostarFixture` depends on `project root path` which binds the instance of protostar to project. At the other hand, the `ProtostarCLI` is also bound to a project after instantiation, so I'll drop the `project` word to be consistent

File: tests/integration/conftest.py (hunk 1 above)
Comment (kasperski95) [reply tests/integration/conftest.py]: Replaced

File: tests/integration/conftest.py (hunk 3 above)
Comment (kasperski95) [reply tests/integration/conftest.py]: Removed the concept of `ProtostarProject` but left the name of this function, because `create_protostar` would be more misleading.

**Review by mkaput — APPROVED** [review]

File: tests/integration/conftest.py [hunk 4]
```diff
@@ -152,20 +154,31 @@ async def run_cairo_test_runner(
...
+    tmp_path_factory: TempPathFactory,
     signing_credentials: Tuple[str, str],
-) -> ProtostarFixture:
-    return build_protostar_fixture(
-        mocker=session_mocker,
-        project_root_path=protostar_project_root_path,
-        signing_credentials=signing_credentials,
-    )
+):
+    @contextmanager
```
Comment (Arcticae) [inline tests/integration/conftest.py:168]: does this have to be context manager? From what i know, we're using i.e. devnet in similar manner and it doesn't have to be used with `with` keyword

File: tests/integration/deploying/deploy_command_test.py [hunk 5]
```diff
@@ -1,34 +1,32 @@
...
+from tests.integration.conftest import CreateProtostarProjectFixture
+from tests.integration.protostar_fixture import ProtostarFixture
+
+
+@pytest.fixture(name="protostar", scope="module")
+def protostar_fixture(create_protostar_project: CreateProtostarProjectFixture):
+    with create_protostar_project() as protostar:
+        protostar.create_files({"./src/main.cairo": CONTRACT_WITH_CONSTRUCTOR})
+        protostar.build_sync()
+        yield protostar
```
Comment (Arcticae) [inline tests/integration/deploying/deploy_command_test.py:15]: this is duplicated code

Comment (Arcticae) [reply tests/integration/deploying/deploy_command_test.py:15]: can we make a few of those, just with different scopes ?

Comment (Arcticae) [reply tests/integration/deploying/deploy_command_test.py:15]: this would make things more clear and re-usable

**Review by Arcticae — CHANGES_REQUESTED** [review]

File: tests/integration/deploying/deploy_command_test.py (hunk 5 above)
Comment (kasperski95) [reply tests/integration/deploying/deploy_command_test.py:15]: > this would make things more clear and re-usable

IMO the opposite would happen.
- This approach allows modifying project files before building the project, and such modification usually needs to be done per test suite / test case.
- The test suite would be harder to understand because it would operate on an environment defined in a different file or files.

Comment (Arcticae) [reply tests/integration/deploying/deploy_command_test.py:15]: >  would operate on an environment defined in a different file or files.
that's usually how our fixtures work, and in terms of reusability it's totally fine. if anything, it makes code cleaner because we don't need to repeat ourselves. 
> such modification usually needs to be done per test suite / test case.
That's usually only adding some test files, or a tree structure right? We could parametrize the fixture with some predefined sources which would be copied to the tree.

File: tests/integration/conftest.py (hunk 4 above)
Comment (kasperski95) [reply tests/integration/conftest.py:168]: Unlike the devnet fixture, the side effect that needs to be reverted is caused when the function is called. There's no point of cleaning up if the function is never called.

Comment (Arcticae) [reply tests/integration/conftest.py:168]: got it

File: tests/integration/deploying/deploy_command_test.py (hunk 5 above)
Comment (Arcticae) [reply tests/integration/deploying/deploy_command_test.py:15]: i can see this kind of code littering our tests all over - that's what i'm afraid of

Comment (kasperski95) [reply tests/integration/deploying/deploy_command_test.py:15]: > that's usually how our fixtures work, and in terms of reusability it's totally fine. if anything, it makes code cleaner because we don't need to repeat ourselves.

It makes the code cleaner visually, but much more obscure logically and harder to maintain.

- colocation https://kentcdodds.com/blog/colocation#the-principle
- the principle of proximity https://kula.blog/posts/proximity_principle/

> That's usually only adding some test files, or a tree structure right? We could parametrize the fixture with some predefined sources which would be copied to the tree.

Not always. Sometimes there's a need to build the project after adding files (migration tests), sometimes there's no such need (`formatter_test.py`). You would have to parametrize the fixture per test case, which would cause duplication (consider `migrator_deploy_contract_cheatcode_test.py`).

Comment (Arcticae) [reply tests/integration/deploying/deploy_command_test.py:15]: This approach won't scale well that's all im saying

**Comment by Arcticae** [comment]
btw i just had a thought about `create_protostar_project` - is there a case where we don't actually create the project? (don't call the function). Each time we want the cleanup logic, so having it as a context manager is kind of weird

**Comment by kasperski95** [comment]
> btw i just had a thought about create_protostar_project - is there a case where we don't actually create the project? (don't call the function). Each time we want the cleanup logic, so having it as a context manager is kind of weird

There's no such case. However, we need to control project scope (function/module). I could add and use two fixtures `protostar_project: ProtostarFixture` and `protostar_project_module: ProtostarFixture`.

@mkaput @Arcticae WDYT?

**Comment by mkaput** [comment]
> > btw i just had a thought about create_protostar_project - is there a case where we don't actually create the project? (don't call the function). Each time we want the cleanup logic, so having it as a context manager is kind of weird
> 
> There's no such case. However, we need to control project scope (function/module). I could add and use two fixtures `protostar_project: ProtostarFixture` and `protostar_project_module: ProtostarFixture`.
> 
> @mkaput @Arcticae WDYT?

I don't like the fact that some tests will use one name while others another.

**Comment by kasperski95** [comment]
> I don't like the fact that some tests will use one name while others another.

So we have to use factory fixture (`create_protostar_project`) as it is now, similarly to `tmp_path_factory`.

The question is where to put the cleanup code. In the fixture or in the `contextmanager`. IMO `create_protostar_project` creates mess (changing cwd) so it should be responsible for cleaning this up.

**Comment by mkaput** [comment]
Yeah, at this point I also lean towards using context managers here.

File: tests/integration/deploying/deploy_command_test.py (hunk 5 above)
Comment (kasperski95) [reply tests/integration/deploying/deploy_command_test.py:15]: I'd say the opposite. It gives flexibility to adapt environment for test suite needs. Without that, test cases would have to adapt to the environment, which would result in lower quality tests. Fixture environments would become very rigid, because any change could break tests.

File: tests/integration/protostar_fixture.py [hunk 6]
```diff
@@ -10,30 +13,20 @@
...
-    MigrateCommand,
-)
-from protostar.commands.init.project_creator.new_project_creator import (
-    NewProjectCreator,
-)
+from protostar.commands import (BuildCommand, DeclareCommand, FormatCommand,
+                                InitCommand, MigrateCommand)
+from protostar.commands.deploy.deploy_command import DeployCommand
+from protostar.commands.init.project_creator.new_project_creator import \
+    NewProjectCreator
```
[the remark this answers is not in the document]
Comment (kasperski95) [reply tests/integration/protostar_fixture.py]: done and configured `isort` to use black profile

File: pyproject.toml [hunk 7]
```diff
@@ -115,6 +115,8 @@ target-version = [
 
 [tool.isort]
 lines_between_sections = 1
+profile = "black"
```
Comment (mkaput) [inline pyproject.toml:118]: could this change be cherry picked to separate PR? :)

**Review by Arcticae — APPROVED** [review]

**Review by mkaput — APPROVED** [review]
